### ISSUE TITLE
Users not logged out when used SSO

### ISSUE THREAD

**Issue by dmatej** [payara/Payara#744]
It can be simulated even in single instance. We have an ear with many war files, SSO turned on. User is logged in, works with one module in one Firefox tab. Then opens another tab and browses to another module. Then logs out in the first tab - it seems everything is alright, but he can still work in the second tab. 

I have turned the monitoring on Web Container - I use only one user and two tabs in one browser window, but now I got 38 sessions open and only 5 invalidated.

I have also enabled logging on javax.enterprise.web (FINEST), but it is also buggy (placeholders are not replaced by values) and suprisingly I got also NPE. It seems the SSO entries are not removed ...? But I don't see why?

The class is GlassFishSingleSignOn ... the NPE is really not a cause of the problem, but it's 
`[2016-04-06T23:01:56.257+0200] [Payara 4.1] [SEVERE] [AS-WEB-CORE-00037] [javax.enterprise.web.core] [tid: _ThreadID=26 _ThreadName=http-listener-1(2)] [timeMillis: 1459976516257] [levelValue: 1000] [[
  An exception or error occurred in the container during the request processing
java.lang.NullPointerException
        at com.sun.enterprise.security.web.GlassFishSingleSignOn.invoke(GlassFishSingleSignOn.java:488)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:702)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:413)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:283)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:283)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:200)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:132)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:111)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:536)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:591)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:571)
        at java.lang.Thread.run(Thread.java:745)
`

**Comment by dmatej** [comment]
I expected that if user calls session.invalidate, all sessions in SSO will be invalidated. But it seems it is implemented to expire SSO after all sessions are invalidated/expired. So finally there must be some storage in application that invalidates all sessions as we need? Or it can be configured somehow?
But there is still the NPE ...

**Comment by dmatej** [comment]
This is from out application log - it seems the browser steals SSO session that should be already invalidated. JSESSION id is new, but JSESSIONIDSSO should not exist.

Cookies:
  JSESSIONID=dde25921f1764468028b2aab03d0; maxAge=-1
  ROUTEID=.; maxAge=-1
  JSESSIONIDSSO=D7421B74884517E4D22A40D7402C3880; maxAge=-1
Session: 
  isNew=true
  id=dde895a9c8c7cf6e11406544b322

It seems that request from one module uses JSESSIONID from another module, where user logged out ... and is successful even if the session is not valid any more.
I don't know if I can prepare some tests for that - I can't still imagine what is going on ...

**Comment by dmatej** [comment]
When I call session.invalidate() the deregister method in GlassfishSingleSignOn is not called. It seems it is only called on expiration from several places, but never on logout ...?!

**Comment by dmatej** [comment]
Finally it works ... the problem is in some inconsistent behavior of our application - I have fixed it now. Sessions leak probably due to some combination of the following factors:

1. Overlapping war contexts (/, /sup, /plu, ...).
2. Using dispatcher.forward in filter to redirect to login page in another module.
3. session.invalidate in another module (logout).

At first I have fixed the problem when I implemented some internal sessionId, ssoSessionId and Session manager that registered any new session and it's SSO ID, and it invalidated all session instances mapped to same SSO ID. It was an ugly hack, but worked; it would leak on expiration, but helped to understand :)

Finally I have refactored our application (EAR):
- login/logout is in one module
- any unlogged request causes HTTP redirect to login page in that module (no forward)

session.invalidate now invalidates all sessions under the same SsoEntry as expected and even without that hack.

I don't know if the issue should be closed, because the implementation is not robust enough ... but it someone should write tests atd first, because this is not simple ...

**Comment by mikecroft** [comment]
Good news - do you have some example code (or a description of how your application uses SSO) that we could use to create some tests? I can raise this as an enhancement then and see if we can iron out any kinks in the implementation

**Comment by dmatej** [comment]
I'm afraid it will take more time than I got now, but I will try in several weeks to prepare something ...

**Comment by mikecroft** [comment]
That would be a great help, thank you!

**Comment by mikecroft** [comment]
Hi @dmatej - did you get time to put together a test case for this?

**Comment by dmatej** [comment]
Aha, I have forgotten that. Finally, the problem was loadbalancer and replication in cluster and maybe even between modules (speculation) - since I have refactored application as I wrote earlier AND configured the loadbalancer to hold session for one minute on same instance, all problems vanished.
I can't imagine how to create some repeatable test case because the problem depends on thread scheduling and it's a bit of random ...

And one more problem that was fixed and that may be connected with that - JSP files with or without a session.
Now it works correctly.
